### ISSUE TITLE
High Code Redundancy in SA-Solver Update Functions and Potential Performance Impact

### ISSUE THREAD

**Issue by MOXOLO** [360CVGroup/RelaCtrl#10]
The `adams_moulton_update` and `adams_moulton_update_few_steps` functions in the  file have a significant amount of repeated code. This redundancy not only makes the codebase harder to maintain but also potentially impacts performance due to redundant computations. 

#### Steps to Reproduce
1. Open the file.
2. Compare the `adams_moulton_update` function (lines 596 - 634) and the `adams_moulton_update_few_steps` function (lines 690 - 741).
3. Notice that a large portion of the code, such as getting the noise schedule, calculating the gradient part, and handling the noise part, is identical.

#### Expected Behavior
The repeated code should be refactored into a common function. This would improve code maintainability, reduce the risk of introducing bugs, and potentially enhance performance by eliminating redundant computations.

#### Actual Behavior
The code contains a large amount of redundancy, which can lead to difficulties in maintenance. For example, if a bug is found in the code related to calculating the gradient part, it needs to be fixed in both functions.

#### Possible Solution
Extract the common parts of the code into a separate function, such as `calculate_update_components`. This function can take the necessary parameters and return the calculated gradient part and noise part. Then, both `adams_moulton_update` and `adams_moulton_update_few_steps` can call this common function to reduce code duplication.

```python
def calculate_update_components(self, order, x, tau, model_prev_list, t_prev_list, t):
    ns = self.noise_schedule
    alpha_t = ns.marginal_alpha(t)
    sigma_t = ns.marginal_std(t)
    lambda_t = ns.marginal_lambda(t)
    alpha_prev = ns.marginal_alpha(t_prev_list[-1])
    sigma_prev = ns.marginal_std(t_prev_list[-1])
    gradient_part = torch.zeros_like(x)
    h = lambda_t - ns.marginal_lambda(t_prev_list[-1])
    t_list = t_prev_list + [t]
    lambda_list = [ns.marginal_lambda(t_list[-(i + 1)]) for i in range(order)]
    gradient_coefficients = self.get_coefficients_fn(order, ns.marginal_lambda(t_prev_list[-1]), lambda_t,
                                                     lambda_list, tau)

    for i in range(order):
        if self.predict_x0:
            gradient_part += (1 + tau ** 2) * sigma_t * torch.exp(- tau ** 2 * lambda_t) * gradient_coefficients[
                i] * model_prev_list[-(i + 1)]
        else:
            gradient_part += -(1 + tau ** 2) * alpha_t * gradient_coefficients[i] * model_prev_list[-(i + 1)]

    if self.predict_x0:
        noise_part = sigma_t * torch.sqrt(1 - torch.exp(-2 * tau ** 2 * h))
    else:
        noise_part = tau * sigma_t * torch.sqrt(torch.exp(2 * h) - 1)

    return gradient_part, noise_part

def adams_moulton_update(self, order, x, tau, model_prev_list, t_prev_list, noise, t):
    assert order in [1, 2, 3, 4], "order of stochastic adams bashforth method is only supported for 1, 2, 3 and 4"
    gradient_part, noise_part = self.calculate_update_components(order, x, tau, model_prev_list, t_prev_list, t)
    noise_part *= noise

    if self.predict_x0:
        x_t = torch.exp(-tau ** 2 * h) * (sigma_t / sigma_prev) * x + gradient_part + noise_part
    else:
        x_t = (alpha_t / alpha_prev) * x + gradient_part + noise_part

    return x_t

def adams_moulton_update_few_steps(self, order, x, tau, model_prev_list, t_prev_list, noise, t):
    assert order in [1, 2, 3, 4], "order of stochastic adams bashforth method is only supported for 1, 2, 3 and 4"
    if order == 2 and self.predict_x0:
        # Do the modification here
        pass
    gradient_part, noise_part = self.calculate_update_components(order, x, tau, model_prev_list, t_prev_list, t)
    noise_part *= noise

    if self.predict_x0:
        x_t = torch.exp(-tau ** 2 * h) * (sigma_t / sigma_prev) * x + gradient_part + noise_part
    else:
        x_t = (alpha_t / alpha_prev) * x + gradient_part + noise_part

    return x_t
